### PR TITLE
Optimizer for memory and storage.

### PR DESCRIPTION
**Pull request by chriseth**
Based on top of PR #1396

### CODE REVIEW

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 1]
```diff
@@ -157,6 +194,60 @@ ExpressionClasses::Id CommonSubexpressionEliminator::initialStackElement(int _st
 	return m_expressionClasses.find(AssemblyItem(dupInstruction(1 - _stackHeight)));
 }
 
+void CommonSubexpressionEliminator::storeInStorage(ExpressionClasses::Id _slot, ExpressionClasses::Id _value)
+{
+	if (m_storageContent.count(_slot) && m_storageContent[_slot] == _value)
+		// do not execute the storage if we know that the value is already there
+		return;
+	m_sequenceNumber ++;
```
Comment (LianaHus) [inline libevmcore/CommonSubexpressionEliminator.cpp]: maybe ++m_sequenceNumber is better? if not then at least without space. you have a couple of them bellow in the code

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 2]
```diff
@@ -179,7 +270,19 @@ bool SemanticInformation::breaksBasicBlock(AssemblyItem const& _item)
...
+		if (_item.instruction() == Instruction::SHA3)
+			return true; //@todo: we have to compare sha3's not based on their memory addresses but on the memory content.
 		InstructionInfo info = instructionInfo(_item.instruction());
+		if (_item.instruction() == Instruction::SSTORE)
+			return false;
+		if (_item.instruction() == Instruction::MSTORE)
+			return false;
+		//@todo: We do not handle the following memory instructions for now:
+		// calldatacopy, codecopy, extcodecopy, mstore8,
+		// msize (not that msize also depends on memory read access)
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: typo: not -> note

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 3]
```diff
@@ -230,26 +343,40 @@ AssemblyItems CSECodeGenerator::generateCode(
 
 	// @todo: provide information about the positions of copies of class elements
 
-	// generate the dependency graph
+	// generate the dependency graph starting from final storage and memory writes and target stack contents
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: Being extremely pedantic with the comments I would say:

// generate the dependency graph starting from final storage, memory writes and target stack contents

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 4]
```diff
@@ -401,22 +591,23 @@ bool CSECodeGenerator::removeStackTopIfPossible()
...
 	m_stack[m_stackHeight] = m_stack[_fromPosition];
 }
 
-void CSECodeGenerator::appendSwapOrRemove(int _fromPosition)
+void CSECodeGenerator::appendOrRemoveSwap(int _fromPosition)
 {
+	assertThrow(_fromPosition != c_invalidPosition, OptimizerException, "");
 	if (_fromPosition == m_stackHeight)
 		return;
 	int nr = m_stackHeight - _fromPosition;
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: maybe use a bit more descriptive name than `nr` here?

File: libevmcore/CommonSubexpressionEliminator.h [hunk 5]
```diff
@@ -163,7 +216,10 @@ class CSECodeGenerator
 	std::map<ExpressionClasses::Id, int> m_classPositions;
 
 	/// The actual eqivalence class items and how to compute them.
-	ExpressionClasses const& m_expressionClasses;
+	ExpressionClasses& m_expressionClasses;
+	/// Keeps information about which storage or memory slots were written to by which operations.
+	/// The operations are sorted ascendingly by sequence number.
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.h]:
```
The operations are sorted by sequence number in ascending order
```

File: libevmcore/CommonSubexpressionEliminator.cpp (hunk 3 above)
Comment (chriseth) [inline libevmcore/CommonSubexpressionEliminator.cpp]: Hm, but it should say "final storage _writes_, final memory writes and target stack contents".

**Pull request merged**